Fix error with transfer-slack and GTFS minTransferTime

t2gran opened the pull request:
### Summary

The current version of OTP2 add transfer-slack to GTFS `minTransferTime` in the OptimizedTransfer service, but not in Raptor. This causes the Optimized transfer to fail, trying to make a transfer witch is no longer possible because of the "extra" transfer-slack added. 

### Issue

closes #4114


### Unit tests

- [ ] TODO

### Code style
✅ 


### Documentation

No do adde

### Changelog

The [changelog file](https://github.com/opentripplanner/OpenTripPlanner/blob/dev-2.x/docs/Changelog.md)
is generated from the pull-request title, make sure the title describe the feature or issue fixed.
To exclude the PR from the changelog add `[changelog skip]` in the title.

hannesj commented:
Shouldn't this feature be behind the feature flag `OTPFeature.MinimumTransferTimeIsDefinitive` in both places? So that the slack is added by default in both places

leonardehrenfried commented on src/main/java/org/opentripplanner/routing/algorithm/transferoptimization/services/TransferGenerator.java:
```diff
@@ -179,33 +179,49 @@ private Collection<? extends TripToTripTransfer<T>> findStandardTransfers(TripSt
+    );
+  }
+
+  private int calcRegularTransferDusrationIncSlack(int transferDurationInSeconds) {
```
There is a typo in the method names.

leonardehrenfried commented on src/main/java/org/opentripplanner/routing/algorithm/transferoptimization/services/TransferGenerator.java:
```diff
@@ -179,33 +179,49 @@ private Collection<? extends TripToTripTransfer<T>> findStandardTransfers(TripSt
+      return calcRegularTransferDusrationIncSlack(transferDurationInSeconds);
     }
+    throw new IllegalStateException(
+      "The constrained transfer is not correct processed in the optimized transfer service. " +
```
```suggestion
      "The constrained transfer is not correctly processed in the optimized transfer service. " +
```

leonardehrenfried commented:
> Shouldn't this feature be behind the feature flag `OTPFeature.MinimumTransferTimeIsDefinitive` in both places? So that the slack is added by default in both places, and only overridden by the minimum transfer time only when the feature is enabled in OTP config.

I must admit that it's not clear to me where exactly the Slack is added in the main routing.

https://github.com/opentripplanner/OpenTripPlanner/blob/8ca9f2ab81acf8c46976a76f9b52affd4baa5ce9/src/main/java/org/opentripplanner/routing/algorithm/raptoradapter/transit/constrainedtransfer/ConstrainedBoardingSearch.java#L169-L187 doesn't do anything about slack - it must happen elsewhere.

leonardehrenfried reviewed (changes requested):
I've made a few suggestions.

optionsome commented:
The summary states that this fixes https://github.com/opentripplanner/OpenTripPlanner/issues/4114 but I think this actually fixes a different issue.

t2gran commented:
This should be ready for a final review, I added the transfer-slack to minTransferTime. I have refactored so the Raptor and OptimizeTransfer both uses the same code. I also added more tests.

MikaelJarfors commented on src/main/java/org/opentripplanner/model/transfer/TransferConstraint.java:
```diff
@@ -196,6 +199,48 @@ public boolean isMinTransferTimeSet() {
+    // Ignore transfer, board and alight slack for min-transfer-time
+    if (isMinTransferTimeSet()) {
+      int minTransferTime = getMinTransferTime() + transferSlack;
+      if (direction.isForward()) {
+        int minTransferBoardTime = sourceTransitArrivalTime + minTransferTime;
+        return OTPFeature.MinimumTransferTimeIsDefinitive.isOn()
+          ? minTransferBoardTime
+          : Math.max(minTransferBoardTime, calcRegularTransferTargetTime.getAsInt());
+      } else {
+        int minTransferBoardTime = sourceTransitArrivalTime - minTransferTime;
+        return OTPFeature.MinimumTransferTimeIsDefinitive.isOn()
+          ? minTransferBoardTime
+          : Math.min(minTransferBoardTime, calcRegularTransferTargetTime.getAsInt());
+      }
```
This Return statement and the one below, could possibly be moved outside the if else?
```suggestion
     if (direction.isForward()) {
        int minTransferBoardTime = sourceTransitArrivalTime + minTransferTime;
      } else {
        int minTransferBoardTime = sourceTransitArrivalTime - minTransferTime;
      }
      return OTPFeature.MinimumTransferTimeIsDefinitive.isOn()
          ? minTransferBoardTime
          : Math.max(minTransferBoardTime, calcRegularTransferTargetTime.getAsInt());
```

t2gran replied:
No, the Math calls min/max depending on the search direction.
